spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/eventhub/AbstractEventHubTemplate.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See LICENSE in the project root for
+ * license information.
+ */
+
+package com.microsoft.azure.spring.integration.eventhub;
+
+import com.google.common.base.Strings;
+import com.microsoft.azure.eventhubs.EventData;
+import com.microsoft.azure.eventhubs.EventHubClient;
+import com.microsoft.azure.eventhubs.EventPosition;
+import com.microsoft.azure.eventprocessorhost.EventProcessorHost;
+import com.microsoft.azure.eventprocessorhost.EventProcessorOptions;
+import com.microsoft.azure.spring.cloud.context.core.util.Tuple;
+import com.microsoft.azure.spring.integration.core.api.CheckpointMode;
+import com.microsoft.azure.spring.integration.core.api.PartitionSupplier;
+import com.microsoft.azure.spring.integration.core.api.StartPosition;
+import com.microsoft.azure.spring.integration.eventhub.converter.EventHubMessageConverter;
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.Setter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.messaging.Message;
+import org.springframework.util.Assert;
+
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Abstract base implementation of event hub template.
+ *
+ * <p>
+ * The main event hub component for sending to and consuming from event hub
+ *
+ * @author Warren Zhu
+ */
+public class AbstractEventHubTemplate {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractEventHubTemplate.class);
+    private final EventHubClientFactory clientFactory;
+
+    @Getter
+    @Setter
+    private EventHubMessageConverter messageConverter = new EventHubMessageConverter();
+
+    @Getter
+    @Setter
+    private StartPosition startPosition = StartPosition.LATEST;
+
+    @Setter
+    @Getter
+    private CheckpointMode checkpointMode = CheckpointMode.BATCH;
+
+    AbstractEventHubTemplate(EventHubClientFactory clientFactory) {
+        this.clientFactory = clientFactory;
+    }
+
+    private static EventProcessorOptions buildEventProcessorOptions(StartPosition startPosition) {
+        EventProcessorOptions options = EventProcessorOptions.getDefaultOptions();
+
+        if (startPosition == StartPosition.EARLISET) {
```

Comment:
suggest to use enum with `Function` here.

Reply:
what do you mean by this?